### PR TITLE
Val - Edges - MediaRanker

### PR DESCRIPTION
**Pull request by valgidzi**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I didn't include a custom model method in my final submission. At one point, I had sorting and spotlight methods in the Work model, but when I deployed to Heroku, it became clear that this wasn't a good choice. Cool things I learned about models/ActiveRecord - using scope in validations and using counter_cache in relations.
Describe how you approached testing that model method. What edge cases did you come up with? | I didn't have a custom method to test, but the closest I came to testing an edge case with validations and relations was testing that a Work would respond to the .votes_count method. I was unsure how to get the test database to mimic the votes_count column I'd added to the database, but I was still able to test that it's a valid method using 'must_respond_to'.
What are `session` and `flash`? What is the difference between them? | Session and flash are both hash-like objects that Rails uses to store data for longer than a single request-response cycle. Session stores data until the browser is closed, but flash only stores the data for one extra request-response cycle. As we've used them so far, session stores user login data and flash stores messages to help the user utilize the site's functionality.
Describe a controller filter you wrote. | I wrote two controller filters - one that finds the logged in user runs before every action, the other checks whether or not a user is logged in before the upvote action (I only use it in the WorksController, but I included it in the ApplicationController because it seemed like something that would be useful for other actions.)
What was one thing that you gained more clarity on through this assignment? | I gained much more clarity on dealing with the database and ActiveRecord - how to query, what can really be accomplished with different validations and relations, and I'm slightly less afraid of it.
What is the Heroku URL of your deployed application? | https://media-ranker-val.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | Perhaps I didn't understand the concepts as well as I thought I did before starting the project, or maybe I missed some simpler ways to accomplish things, but to me this felt a little bit closer to a stage 3 project than a stage 2 project - especially when compared to RideShare Rails. I think that either giving a little more direction or bumping it up to a proper stage 3 project would be helpful in the future. I learned a ton, though.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    if @user.nil?
+      head :not_found
+    end
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def create
```
Comment (droberts-sea) [inline app/controllers/users_controller.rb:18]: How are `new` and `create` here different than in the `SessionsController`? Do you need both for this site?

File: app/controllers/pages_controller.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class PagesController < ApplicationController
+  def index
+    @works = Work.order(votes_count: :desc)
+    @books = @works.where(category: 'book')
+    @albums = @works.where(category: 'album')
```
Comment (droberts-sea) [inline app/controllers/pages_controller.rb:5]: While I still think `top_ten` or `spotlight` would make a good model method, using the counter cache makes this controller code pretty clean - you're definitely still offloading a lot of the work to the model.

It sounds like you've identified this, but the interesting piece you're missing by not having this logic isolated is testing. For example, for the spotlight, I would wonder:

- What happens if there are no works?
- What happens if there are works but no votes?
- What happens if two works have the same number of votes?

Similarly for the top ten, I would ask:

- What if there are no works of that category?
- What if there are less than 10 works?
- What if there's a tie for last place, e.g. works 9, 10 and 11 all have 0 votes?

File: app/views/pages/_top_works_list.html.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<h2 class='top-ten-header'><%= list_name %></h2>
+<ul class='top-ten'>
+
+  <% model.first(10).each do |work| %>
```
Comment (droberts-sea) [inline app/views/pages/_top_works_list.html.erb:4]: Good work using a partial to DRY this up. However, I would probably not call `.first(10)` here, but rather in the controller (or better a model method).

File: test/models/vote_test.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  before do
+    @vote = votes(:vote)
+  end
+
+  describe 'validations' do
+    it "is valid with work and user" do
+      expect(@vote.valid?).must_equal true
+    end
+    it "is invalid if user has already voted for work" do
+      @error_vote = Vote.new(
```
Comment (droberts-sea) [inline test/models/vote_test.rb:13]: This is a solid start, but there's a couple other interesting cases that I would like to see here:

- A user can have votes for two different works
- A work can have votes from two different users

File: test/models/work_test.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      expect(@work.votes).must_respond_to :each
+      expect(@work.votes.length).must_equal 4
+    end
+    it 'responds to upvoted_users method' do
+      expect(@work.upvoted_users).must_respond_to :each
+      expect(@work.upvoted_users.length).must_equal 4
+    end
+    it 'responds to votes_count method' do
+      expect(@work).must_respond_to :votes_count
+    end
```
Comment (droberts-sea) [inline test/models/work_test.rb:52]: Since Rails is handling the logic here, this test is probably sufficient. However, if you did want to test it, you could create a new work, create a bunch of votes for it in a `.times` loop, and then check that `votes_count` gives back the right number.

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Comprehension questions	| yes - I am surprised that your model methods didn't work in deployment, I would be interested to take a look at that.
**General** |
Rails fundamentals (RESTful routing, use of named paths) | looks like there may be some extra routes
Views are well-organized (DRY, use of semantic HTML, use of partials) | yes
Errors are reported to the user | yes
Business logic lives in the models | some - see inline
Models are thoroughly tested, including relations, validations and any custom logic | yes, but see inline
**Wave 1 - Media** |
Splash page shows the three media categories | yes
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | yes
Individual user pages and the user list are present | yes
**Optional - Styling** |
Bootstrap is used appropriately | yes
Look and feel is similar to the original | yes
**Overall** | Good work overall. This submission matches the look and feel of the original very closely, and it's clear to me that the learning goals for this assignment were met. I do want to make sure you're getting practice working with and testing custom model methods - MediaRanker Revisited is a good example of how this could be approached. Other than that I'm happy with this submission - keep up the hard work!
